overlord/devicestate/users.go

```diff
@@ -0,0 +1,330 @@
+	"github.com/snapcore/snapd/overlord/assertstate"
+	"github.com/snapcore/snapd/overlord/auth"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/release"
+	"github.com/snapcore/snapd/strutil"
+)
+
+var (
+	osutilAddUser = osutil.AddUser
+	osutilDelUser = osutil.DelUser
+	userLookup    = user.Lookup
+)
+
+// UserError is returned when invalid or insufficient data is supplied,
+// or if a user-assertion is not found.
+type UserError struct {
+	Err error
+}
+
+func (e *UserError) Error() string {
+	return e.Err.Error()
+}
+
+// CreatedUser is the returned result which wraps resulting data
+// from a create user operation.
+type CreatedUser struct {
+	Username string
+	SSHKeys  []string
+}
+
+// CreateUser creates a Linux user based on the specified email.
+// The username and public ssh keys for the created account are
+// determined from Ubuntu store based on the email.
+func CreateUser(st *state.State, sudoer bool, email string) (*CreatedUser, error) {
+	if email == "" {
+		return nil, &UserError{Err: fmt.Errorf("cannot create user: 'email' field is empty")}
+	}
+
+	storeService := snapstate.Store(st, nil)
+	username, opts, err := getUserDetailsFromStore(st, storeService, email)
+	if err != nil {
+		return nil, &UserError{Err: err}
+	}
+
+	opts.Sudoer = sudoer
+	return addUser(st, username, email, opts)
+}
+
+// CreateKnownUsers creates known users. The user details are fetched
+// from existing system user assertions.
+// If no email is passed, all known users will be created based on valid system user assertions.
+// If an email is passed, only the corresponding system user assertion is used.
+func CreateKnownUsers(st *state.State, sudoer bool, email string) ([]*CreatedUser, error) {
+	model, err := findModel(st)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create user: cannot get model assertion: %v", err)
+	}
+
+	serial, err := findSerial(st, nil)
+	if err != nil && !errors.Is(err, state.ErrNoState) {
+		return nil, fmt.Errorf("cannot create user: cannot get serial: %v", err)
+	}
+
+	db := assertstate.DB(st)
+	if email == "" {
+		return createAllKnownSystemUsers(st, db, model, serial, sudoer)
+	}
+
+	username, opts, err := getUserDetailsFromAssertion(db, model, serial, email)
+	if err != nil {
+		return nil, &UserError{Err: err}
+	}
+
+	opts.Sudoer = sudoer
+	createdUser, err := addUser(st, username, email, opts)
+	if err != nil {
+		return nil, err
+	}
+	return []*CreatedUser{createdUser}, nil
+}
+
+// RemoveUser removes linux user account of passed username.
+func RemoveUser(st *state.State, username string) (*auth.UserState, error) {
+	// TODO: allow to remove user entries by email as well
+
+	// catch silly errors
+	if username == "" {
+		return nil, &UserError{Err: fmt.Errorf("need a username to remove")}
+	}
+
+	// check the user is known to snapd
+	_, err := auth.UserByUsername(st, username)
+	if err == auth.ErrInvalidUser {
```

Comment:
nitpick: `errors.Is` might be better here.

Reply:
Thanks!